### PR TITLE
Autoregister blocks

### LINKED ISSUE #471: Autoregister blocks
Today you have to register every newly created block under `sonata_block.blocks` (https://github.com/sonata-project/SonataBlockBundle/blob/3.x/src/DependencyInjection/Configuration.php#L127).

This is really annoying. So, what do you think about an autoregistration. Is there any case where you don't want to register a block that is tagged with `sonata.block`?

The same logic could be applied to breadcrumbs tagged with `sonata.breadcrumb`.

### PR DESCRIPTION
**Pull request by core23**
I am targeting this branch, because this feature is BC

Closes #471

## Changelog

```markdown
### Added
- All blocks containing `sonata.block` will be auto registered
```

## To do
- [x] Update the tests
- [x] Update the documentation

## Subject

You don't have to add all blocks to the config anymore, but you can to override the auto config.

### CODE REVIEW

File: src/Util/StringUtil.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\BlockBundle\Util;
+
+class StringUtil
```
Comment (greg0ire) [inline src/Util/StringUtil.php]: How about `Sonata\BlockBundle\BlockName\ConvertFromFqcn::__invoke` instead, with that class final? That kind of "util" class often ends up growing to the infinite, and being cluttered with unrelated methods whose which have only the `string` => `string` signature in common.

File: src/DependencyInjection/Compiler/TweakCompilerPass.php [hunk 2]
```diff
@@ -99,4 +118,18 @@ public function applyContext(ContainerBuilder $container)
             }
         }
     }
+
+    /**
+     * @return array
```
Comment (greg0ire) [inline src/DependencyInjection/Compiler/TweakCompilerPass.php]: Please make this more precise

File: tests/Util/StringUtilTest.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+use Sonata\BlockBundle\Util\StringUtil;
+
+class StringUtilTest extends TestCase
+{
+    /**
+     * @dataProvider fqcnToBlockNameProvider
+     */
+    public function testFqcnToBlockName($fqcn, $expectedBlockName)
+    {
+        $blockName = StringUtil::fqcnToBlockName($fqcn);
```
Comment (greg0ire) [inline tests/Util/StringUtilTest.php]: No need for that variable IMO

File: src/Util/StringUtil.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    private function __construct()
+    {
+    }
+
+    /**
+     * Converts a fully-qualified class name to a block name.
+     *
+     * @param string $fqcn The fully-qualified class name
+     *
+     * @return string|null The block name or null if not a valid FQCN
```
Comment (greg0ire) [inline src/Util/StringUtil.php]: Why not throw on invalid FQCN?

File: UPGRADE-3.x.md [hunk 5]
```diff
@@ -1,6 +1,15 @@
 UPGRADE 3.x
 ===========
 
+UPGRADE FROM 3.12 to 3.x
+=======================
+
+## Autoregister blocks
+
+All blocks tagged with `sonata.block` will now be autoregistered to the `sonata.block.loader.service`.
```
Comment (greg0ire) [inline UPGRADE-3.x.md]: "to" => "into"

File: src/Naming/ConvertFromFqcn.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+namespace Sonata\BlockBundle\Naming;
+
+use Sonata\BlockBundle\Naming\Exception\NamingException;
+
+/**
+ * Converts a fully-qualified class name to a block name.
+ *
+ * @author Christian Gripp <mail@core23.de>
+ */
+class ConvertFromFqcn
```
Comment (greg0ire) [inline src/Naming/ConvertFromFqcn.php]: Please make it `final`

File: src/Naming/ConvertFromFqcn.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+{
+    /**
+     * @param string $fqcn The fully-qualified class name
+     *
+     * @return string The block name or null if not a valid FQCN
```
Comment (greg0ire) [inline src/Naming/ConvertFromFqcn.php]: This comment is no longer accurate

File: docs/reference/installation.rst [hunk 8]
```diff
@@ -68,20 +68,6 @@ To use the ``BlockBundle``, add the following lines to your application configur
         sonata_block:
             default_contexts: [sonata_page_bundle]
             blocks:
-                sonata.admin.block.admin_list:
```
Comment (kunicmarko20) [inline docs/reference/installation.rst]: https://github.com/symfony/recipes-contrib/blob/master/sonata-project/admin-bundle/3.31/config/packages/sonata_admin.yaml

should probably be changed here also

File: src/Naming/ConvertFromFqcn.php [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+     * @return string The block name
+     */
+    public function __invoke($fqcn)
+    {
+        // Non-greedy ("+?") to match "service" suffix, if present
+        if (preg_match('~([^\\\\]+?)(service)?$~i', $fqcn, $matches)) {
+            return strtolower(preg_replace(['/([A-Z]+)([A-Z][a-z])/', '/([a-z\d])([A-Z])/'], ['\\1_\\2', '\\1_\\2'], $matches[1]));
+        }
+
+        throw new NamingException(sprintf('The name "%s" is not a valid FCQN', $fqcn));
```
Comment (OskarStark) [inline src/Naming/ConvertFromFqcn.php]: You didn’t catch it somewhere, so why don’t use InvalidArgumentException?

If not InvalidArgumentException, maybe InvalidBlockIdException ??

File: tests/Naming/ConvertFromFqcnTest.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\BlockBundle\Tests\Naming;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\BlockBundle\Naming\ConvertFromFqcn;
+
+class ConvertFromFqcnTest extends TestCase
```
Comment (OskarStark) [inline tests/Naming/ConvertFromFqcnTest.php]: Please add a testcase for the exception case

File: src/Naming/ConvertFromFqcn.php (hunk 9 above)
Comment (core23) [reply src/Naming/ConvertFromFqcn.php]: The service has nothing todo with block ids. It's just about naming blocks.

File: src/DependencyInjection/Compiler/TweakCompilerPass.php [hunk 11]
```diff
@@ -116,4 +136,18 @@ private function replaceBlockName(ContainerBuilder $container, Definition $defin
             );
         }
     }
+
+    /**
+     * @return string[]
+     */
+    private function getContextFromTags(array $tags)
+    {
+        return array_filter(array_map(function ($attribute) {
```
Comment (greg0ire) [inline src/DependencyInjection/Compiler/TweakCompilerPass.php]: Missing type hints

File: src/DependencyInjection/Compiler/TweakCompilerPass.php [hunk 12]
```diff
@@ -116,4 +136,18 @@ private function replaceBlockName(ContainerBuilder $container, Definition $defin
             );
         }
     }
+
+    /**
+     * @return string[]
```
Comment (greg0ire) [inline src/DependencyInjection/Compiler/TweakCompilerPass.php]: Are you sure? When reading the code, I'm not under the impression that `$attribute` is a string.

Comment (core23) [reply src/DependencyInjection/Compiler/TweakCompilerPass.php]: Yes, it's a string, if you declate a `<tag ... context="foo">` inside your `services.xml`

Comment (core23) [reply src/DependencyInjection/Compiler/TweakCompilerPass.php]: No, the method should return just a string array

Comment (greg0ire) [reply src/DependencyInjection/Compiler/TweakCompilerPass.php]: Ooh right my bad, I thought this was an `@param` phpdoc! Can you please add it?

Comment (greg0ire) [reply src/DependencyInjection/Compiler/TweakCompilerPass.php]:
```suggestion
     * @param string[][]
     *
     * @return string[]
```

**Comment by greg0ire** [comment]
Thanks @core23 !

**Pull request merged**